Initial draft of Python 'Security Policy' (#1804)
python/devguide: security-and-threat-model/index.rst

=========================
Security and threat model
=========================

The majority of Python Security Response Team (PSRT)
members are volunteers. Therefore, you must respect this volunteered time
by following this security policy. Repeated failure to
respect the security policy will result in future reports

> being rejected or being banned from the `python` GitHub organization, regardless of technical merit.

Reviewer: ```suggestion
being rejected or being banned from the ``python`` GitHub organization, regardless of technical merit.
```
Also, please wrap lines.

Author: Done, and wrapped lines in upcoming commit too.

triggerable with data inputs that are reasonably sized for the use-case.
Availability vulnerabilities must also demonstrate an "upward" change in posture
for the attacker, rather than a "lateral" change in posture.
This is to avoid handling performance improvements as security vulnerabilities.

Vulnerabilities in dependencies of Python (such as zlib, Tcl/Tk, or OpenSSL)
are not vulnerabilities in Python unless Python's use of the dependency
interferes with secure use of the dependency.
For example, Python is not vulnerable because it bundles a vulnerable
version of zlib, users are expected to upgrade their own dependencies.

The complete threat model for Python and standard library modules
is available in the Threat Model section of the Python Developer Guide.

What versions of Python are accepting reports?
----------------------------------------------

Python accepts security vulnerabilities and will
assign CVE IDs for `supported Python versions`_ that have a status of
`"bugfix" or "security" <python-status>`_. Versions that are not yet
stable (status of `"feature" or "prerelease" <python-status>`_) are not
eligible for CVE IDs. If the vulnerability exclusively exists in

> non-stable versions, then the issue should be handled as a public bug issue.

Reviewer 2: Might go without saying, but maybe we should include something like: "please articulate if the vulnerability is no longer reproducible at HEAD and only exists in specific stable releases, and if so which ones"?

Reviewer 2: Ah, gps said something similar here: https://github.com/python/devguide/pull/1804#discussion_r3268516923

Sometimes features may be marked as
"experimental" in Python, even in a stable Python version.
These features are not eligible for security vulnerabilities.
Instead open a public GitHub issue.

If a vulnerability is platform-dependent, check if the platform is

> `supported per :pep:`11`.

Reviewer: ```suggestion
supported per :pep:`11`.
```

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

Vulnerabilities that exclusively affect unsupported platforms
may not be accepted.

.. _supported Python versions: https://devguide.python.org/versions/

> .. _python-status: https://devguide.python.org/versions/#status-key

Reviewer: Please use `:ref:`s, if these pages/sections ever move it saves the hassle of fixing links.

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

What to include and how to structure a vulnerability report?
------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and

>   a proof-of-concept script that reproduces the issue.

Reviewer 2: Regarding a proof-of-concept script, can we include something like:

> Ideally a proof-of-concept script should exit with code 0 if the vulnerability is not reproducible, and exit with a code of 1 if the vulnerability is reproduced.

Reviewer 2: Also, I would argue that a proof-of-concept script could be made a requirement at this point in time, given the volume we're seeing.

Author: I think we'll keep it optional for now, since another comment in this thread mentions sometimes the POC is so complex that it's not even worth creating and reviewing for us.

=========================
Security and threat model
=========================

> The majority of Python Security Response Team

Reviewer 3: ```suggestion
The majority of Python Security Response Team (PSRT)
```

> members are volunteers and therefore you must respect this volunteered time

Reviewer 4: ```suggestion
members are volunteers. Therefore, you must respect this volunteered time
```

> members are volunteers and therefore you must respect this volunteered time

Reviewer 5: wording suggestion:

"""The majority of the Python Security Response Team (PSRT) members are volunteers. You must respect the time of security responders by following this policy. Repeated failure to ..."""

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

by following this security policy. Repeated failure to
respect the security policy will result in future reports

> being rejected, regardless of technical merit.

Reviewer 4: ```suggestion
being rejected or being banned from the `python` GitHub organization, regardless of technical merit.
```

We should also remind folks that security reports are subject to the PSF Code of Conduct.

Reviewer 4: Oh, I see that there is a section about it below. Perhaps moving it up would be better?

Author: So I can accept your suggestion I've copied your second comment into its own resolvable thread: https://github.com/python/devguide/pull/1804/changes#r3268096748

What types of bugs are vulnerabilities?
---------------------------------------

Not all bugs are vulnerabilities. To avoid causing
duplicate work for PSRT members all potential reports

> must be evaluated against the relevant threat model(s)

Reviewer 3: "(s)" is a little awkawrd
```suggestion
must be evaluated against the relevant threat models
```

prior to being submitted to the PSRT.

Vulnerabilities must be exploitable from code, configurations,
pre-conditions, and deployments that might feasibly exist in

> the real-world. For example, a vulnerability only affecting code

Reviewer 3: ```suggestion
the real world. For example, a vulnerability only affecting code
```

> that does not make sense to write in a production program

Reviewer 3: ```suggestion
that does not make sense in a production program
```

will not be accepted as a vulnerability.

Documented functionality will not be considered a vulnerability.

> For example, ``pickle``, ``marshal``, ``shelve``, ``eval``, and ``exec``

Reviewer: You can link to them with `:mod:`.

are documented to execute arbitrary Python code that is supplied as data.
The ``ctypes`` module is documented to enable modifying arbitrary locations

> in memory.

Reviewer 3: Adding `:mod:` and rewrapping:
```suggestion
For example, :mod:`pickle`, :mod:`marshal``, :mod:`shelve``, :mod:`eval``,
and :mod:`exec` are documented to execute arbitrary Python code that is supplied
as data. The :mod:`ctypes` module is documented to enable modifying arbitrary locations in memory.
```

Vulnerabilities must not depend on malicious control of:

> * what Python code is executed by the interpreter

Reviewer 6: This probably needs clarifying, since if someone remotely persuades a runtime to execute their arbitrary code, that would be a vulnerability.

Perhaps we just want to say something like this instead of the bullet list?

Vulnerabilities must not depend on malicious control of Python's launch conditions, including command line arguments, environment variables, or modifications to files on the target system. We assume that, at the time Python is executed, the environment is as intended by the legitimate user, and any malicious variation from this  cannot be mitigated by Python itself.

* locations where code is loaded prior to execution (such as current working
  directory, ``PYTHONPATH``)
* configuration files
* environment variables
* command line options
* installed packages or modules
* `.pth files <https://docs.python.org/3/library/site.html>`__
* caches or ``.pyc`` files

Vulnerabilities that affect availability (such as DoS, ReDoS) must be
triggerable with data inputs that are reasonably sized for the use-case.
Availability vulnerabilities must also demonstrate an "upward" change in posture
for the attacker, rather than a "lateral" change in posture.
This is to avoid handling performance improvements as security vulnerabilities.

Vulnerabilities in dependencies of Python (such as zlib, Tcl/Tk, or OpenSSL)
are not vulnerabilities in Python unless Python's usage of the dependency

> interferes with secure usage of the dependency.

Reviewer 3: ```suggestion
are not vulnerabilities in Python unless Python's use of the dependency
interferes with secure use of the dependency.
```

For example, Python is not vulnerable because it bundles a vulnerable

> version of zlib, users are expected to upgrade their own dependencies.

Reviewer 7: Comma splice. Also, is this right? Seems like we often take security upgrades of bundled dependencies such as expat.

Reviewer: Similarly we also bundle the deps on some platforms (Windows), so our sources (`cpython-*-deps`) need to be updated in the event of a security vulnerability.

Reviewer 4: I think the idea is that we do not currently create patch releases when a dependency has a security vulnerability, or mitigate against vulnerable system dependencies.

Reviewer 6: Plus users can't be expected to upgrade a _bundled_ dependency, so the sentence needs a bit of reframing.

"For example, a vulnerability in the bundled copy of zlib in Python is a vulnerability in zlib, not Python."

I don't think we need to mention "users can upgrade their own dependencies". Maybe we need a written policy somewhere else on how we choose to release updates to our binary distros in light of bundled libraries though? Not here I don't think.

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

> The complete threat model for Python and standard library modules,

Reviewer 7: ```suggestion
The complete threat model for Python and standard library modules
```

> is available in the Threat Model section of the Python Developer Guide.

Reviewer 7: Is that a reference to this page?

Reviewer: That page doesn’t exist yet, so let's not reference it.

What versions of Python are accepting reports?
----------------------------------------------

Python accepts security vulnerabilities and will
assign CVE IDs for `supported Python versions`_ that have a status of
`"bugfix" or "security" <python-status>`_. Versions that are not yet
stable (status of `"feature" or "prerelease" <python-status>`_) are not
eligible for CVE IDs. If the vulnerability exclusively exists in

> non-stable versions, then the issue should be handled as a public bug issue.

Reviewer 6: ```suggestion
eligible for CVE IDs. If the vulnerability only exists in prerelease
versions (alphas, betas, RCs), then the issue should be reported as a regular bug.
```

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

Sometimes features may be marked as
"experimental" in Python, even in a stable Python version.
These features are not eligible for security vulnerabilities,

> instead open a public GitHub issue.

Reviewer 7: ```suggestion
These features are not eligible for security vulnerabilities.
Instead open a public GitHub issue.
```

If a vulnerability is platform-dependent, check if the platform is a

> `supported platform per PEP 11 <https://peps.python.org/pep-0011/>`__.

Reviewer 3: ```suggestion
If a vulnerability is platform-dependent, check if the platform is
`supported per :pep:`11`.
```

may not be accepted.

.. _supported Python versions: https://devguide.python.org/versions/
.. _python-status: https://devguide.python.org/versions/#status-key

What to include and how to structure a vulnerability report?
------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and
  a proof-of-concept script that reproduces the issue.
* When reporting large numbers or "batches" of vulnerabilities or
  searching for potential vulnerabilities using an LLM, you as a reporter must
  verify the validity of all reports prior to submission to the PSRT.
  PSRT members WILL NOT spend time confirming the validity of reports, only
  whether a valid bug report is a vulnerability or not.
* Do not include severity or CVSS information in your initial report,
  this information will be determined by the PSRT.

> * Optionally, include a minimal patch with the mitigation for the report.

Reviewer 4: ```suggestion
* Ideally, include a minimal patch with the mitigation for the report.
```

* If the vulnerability only affects certain Python versions, optionally
  include the versions of Python that are affected.
* Reports that do not contain a potential security vulnerability (such as spam
  or requesting compliance or due-diligence work)
  will be discarded without a reply.

How to submit a vulnerability report?
-------------------------------------

Submit all potential security vulnerability reports for CPython
to `GitHub Security Advisories <GHSA>`__

> by `opening a new ticket <GHSA>`__.

Reviewer: This is excessive linking, see https://devguide.python.org/documentation/style-guide/#links.

Reviewer 3: ```suggestion
to GitHub Security Advisories
by `opening a new ticket <GHSA>`__.
```

Do not open a public GitHub issue to report a security vulnerability.

> For all other projects (pip, python.org, tools, etc) or if you're

Reviewer: Please don't use latin abbreviations: https://devguide.python.org/documentation/style-guide/#use-simple-language

Reviewer 3: ```suggestion
For all other projects (such as pip, python.org and tools) or if you're
```

not sure where to send your report, send an email to

> `security at python dot org <mailto:security@python.org>`__.

Reviewer: ```suggestion
`security@python.org <mailto:security@python.org>`__.
```
In case people want to copy paste.

Here's what to expect for how a vulnerability report will be handled:

* Reporter reports the vulnerability privately to the PSRT.
* If the PSRT determines the report isn't a vulnerability, the issue
  can be opened in the public issue tracker.
* If the PSRT determines the report is a vulnerability, the PSRT will
  accept your report and a CVE ID will be assigned by the PSF CNA.
* Once a public pull request containing a fix is merged to CPython,
  the advisory and CVE record will be published with attribution.

For more information about how the PSRT handles vulnerabilities,

> `consult the Python Developer Guide <https://devguide.python.org/developer-workflow/psrt/#triaging-a-vulnerability-report>`__.

Reviewer: This should be a `:ref:`, it is more future proof if the page moves.

PSF Code of Conduct
-------------------

Well-being and safety of the Python Security Response Team members is
prioritized over the technical merit of vulnerability reports.
Despite communications being private, vulnerability reporting is subject
to the `PSF Code of Conduct`_. Violations will be reported to the Code of Conduct
team with undisclosed vulnerability information removed, if applicable.

.. _GHSA: https://github.com/python/cpython/security/advisories/new
.. _PSF Code of Conduct: https://policies.python.org/python.org/code-of-conduct/

CVE Numbering Authority (CNA)
-----------------------------

The Python and pip projects are scoped under the
`Python Software Foundation CVE Numbering Authority <CNA>`__
(CNA). This means you must submit all security
vulnerability reports to the PSRT to receive
a CVE ID for Python or pip. To reach the PSF
CNA contact directly, send an email to

> `cna at python dot org <mailto:cna@python.org>`__.

Reviewer: ```suggestion
`cna@python.org <mailto:cna@python.org>`__.
```

Security and threat model
=========================

The majority of Python Security Response Team (PSRT)
members are volunteers. Therefore, you must respect this volunteered time

> by following this security policy. Repeated failure to

Reviewer 8: Tying "respect their time" to the volunteer aspect feels a bit off to me. Maintainers that have a carve out in their day job for this work deserve the same consideration. 

Could we drop the conditional framing? Maybe something like "PSRT members balance this work against many other responsibilities. Please be thoughtful about the time and attention your report requires."

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed
